src/github/amexboy/modules_loader.clj

```diff
@@ -0,0 +1,52 @@
+(ns github.amexboy.modules-loader
+  (:require [clojure.java.io :as io])
+  (:use [clojure.java.shell :only [sh]]
+        [github.amexboy.git-ops]
+        [github.amexboy.project-gen]))
+
+(defn list-modules [root]
+  "Returns a list of module name in this root"
+  (let [dir (io/file root)]
+    (filter maven-root? (.listFiles dir)))
+  )
+
+(defn pull-module [root module config]
+  "Pull module from github"
+  (when-not (:no-clone config)
+    (git-clone (str (:repository-root config) module) root)))
+
+(defn update-repository [root module config]
+  "Checks if module exists and updates the repository, pull from github if not"
+  (let [module-file (io/file root module)]
+    (if-not (.exists module-file)
+      (pull-module root module config)
+      true)))
+
+(defn create-module [root module branch name opts]
+  (let [src (.getPath (io/file root module))
+        work-tree-file (io/file root (:modules-root opts) name module)
+        work-tree (.getPath work-tree-file)]
+    (if (update-repository root module opts)
+      (do
+        (println module)
+        (git-add-worktree src branch work-tree (:base-branch opts))
+        (cond
+          (maven-root? work-tree-file) (gen-maven-iml work-tree module)
+          (gradle-root? work-tree-file) (gen-gradle-iml work-tree module)
+          (liengen-root? work-tree-file) (gen-liengen-iml work-tree module)
+          )
+        module)
+      false)))
+
+(defn create-project
+  "Creates a new ipr file and loads the module checking out the branch on the modules"
+  [root name branch modules opts]
+  (gen-ipr (io/file root (:modules-root opts) name)
+           name
+           (filter identity (map #(create-module root % branch name opts) modules))
```

Comment:
Use `keep` instead of `(filter identity ...)`. `identity` is almost never useful in production code, though it's very useful at the REPL.

Reply:
`create-module` appears to be writing to disk. Don't do side-effects inside sequence functions (e.g. `map`, `filter`). Do a `doseq`, and _then_ continue other work:

```clj
(doseq [m module]
  (create-module ... m ...))
(gen-ipr ...)
```

If you need to track some state, I recommend using `loop/recur` instead:

```clj
(loop [mods modules
       created #{}]
  (when-some [m (first mods)]
    (if (create-module ...)
      (recur (next mods) (conj created m)
      (recur (next mods) created)))
```

Though, I think that if you iron all of this disparate logic into larger chunks, you'll find you don't need to track quite so much state in the first place. So I would be hesitant to jump right to `loop/recur`.

Reply:
> Use 'keep' instead

huh - this is new trick for me. I've been using `(remove nil? (...))` for a long time. `(filter identity (...))` would have never occurred to me